connection/cluster/ClusterClient.java

```diff
@@ -148,9 +149,13 @@ ClusterSession run(ClusterDatabase.Replica replica) {
         };
     }
 
-    // TODO: this is not good - we should not pass an internal object to be modified outside of this class
-    ConcurrentMap<String, ClusterDatabase> databaseByName() {
-        return clusterDatabases;
+    @Nullable
+    ClusterDatabase databaseGet(String name) {
+        return clusterDatabases.get(name);
+    }
+
+    void databasePut(String name, ClusterDatabase database) {
+        clusterDatabases.put(name, database);
```

Comment:
While this is marginally better than before because of less access to the `HashMap`, I think the overall problem remains - we are still building up the definition of what a `ClusterClient` is by adding stuff to it from outside of the class.

We could solve this issue by adding a callback function to the invocations of `ClusterClient`, say, `onFetchDatabaseReplicas`, but this doesn't feel like the right solution because the actual value of the callback would always be the same - we always want to call `client.databasePut`.

Because of the strong coupling between `ClusterClient` and `FailsafeTask`, and also the awkward positioning of `FailsafeTask` in the `cluster` package (note: every other class in this package starts with the prefix `Cluster`!) I'm starting to feel like `FailsafeTask` should actually be an inner class of `ClusterClient`.

That way, we don't need to make this change at all, because `databaseByName` can be dissolved, as `clusterDatabases` will now be used only within the class itself.